### ISSUE TITLE
Add config option for upgrade polling URL.

### ISSUE THREAD

**Issue by jrbriggs** [Microsoft/VFSForGit#923]
To support decoupling the upgrade of multiple engineering systems, we should move the upgrade source of truth from GH to an anonymously-available http service that returns a json block with current versions for each ring.

**Comment by turbonaitis** [comment]
Will this URL be customizable, so that VFS4G users would be able to have their own update feed (either to control when a new version is released, or to release their own, forked, versions)?

**Comment by jamill** [comment]
@turbonaitis this is something that has been discussed. Do you have any feedback on how / if you might use this functionality? The current plan is that external users would update via the public GitHub feed, but I am looking for feedback on how / if other orgs might want to leverage this feature.

**Comment by turbonaitis** [comment]
So if I'm reading this issue correctly, there will be an HTTP endpoint, that will return a set of different VFS4G versions for different rings. If that looks something like this
```
{
   "stable":{
      "version":"2.2",
      "url":"http://nuget.org/vfs4g/2.2/installer.pkg"
   },
   "beta":{
      "version":"2.3",
      "url":"http://nuget.org/vfs4g/2.3/installer.pkg"
   }
}
```
Then the ability to customize the URL of this service is all we want. 

Now, that I think of it, we also want to consider a security aspect here, so that a malicious actor can't simply update this setting to point to a poisoned feed. Ideally, the packages would be signed and we'd have a public key to verify them. 

Perhaps an alternative would be to specify this URL at compile time and always require it to be an https endpoint. 

Securing a package registry is wide topic, I guess, and is not my area of expertise. 

@Ignas - would you be able to provide some input here?

**Comment by jamill** [comment]
> Now, that I think of it, we also want to consider a security aspect here, so that a malicious actor can't simply update this setting to point to a poisoned feed.

Yes - this is something we have kept in mind with our thread models. One mitigation is that `gvfs.config` requires elevated privileges to edit. We also require the packages to be signed (with the expected certificate).

> Perhaps an alternative would be to specify this URL at compile time and always require it to be an https endpoint.

The current design was this endpoint would return a specific version for a particular org / ring combination, but not have the actual URL for the NuGet feed included. The upgrade logic would then use the version specified by this endpoint to download a specific version of the package from the "known" NuGet feed. 

> require it to be an https endpoint.

This seems like a reasonable requirement. We have been relying on the package verification with a well known cert to ensure the integrity of the downloaded package (and that it was produced by the expected author).

I could put up more details of our current design, if that is something of interest.

/cc @jrbriggs - as I have we have had discussions around several of these topics.

**Comment by turbonaitis** [comment]
> > Now, that I think of it, we also want to consider a security aspect here, so that a malicious actor can't simply update this setting to point to a poisoned feed.
> 
> Yes - this is something we have kept in mind with our thread models. One mitigation is that `gvfs.config` requires elevated privileges to edit. We also require the packages to be signed (with the expected certificate).
If we, or someone else, would be to build from a fork, we'd obviously use a different key pair to sign the resulting package. Is there a way to customize this "expected certificate"

> 
> > Perhaps an alternative would be to specify this URL at compile time and always require it to be an https endpoint.
> 
> The current design was this endpoint would return a specific version for a particular org / ring combination, but not have the actual URL for the NuGet feed included. The upgrade logic would then use the version specified by this endpoint to download a specific version of the package from the "known" NuGet feed.
Is that NuGet feed configurable, or hardcoded? Because again, other organisation might want to download the binaries from their own locations, if they've built them from a fork. Also, I assume any URL would work, not necessarily conforming to NuGet's formatting. 
> I could put up more details of our current design, if that is something of interest.
This would be helpful, thank you.
